### PR TITLE
Added 'is_public' and 'is_open' columns to calendar event.

### LINKED ISSUE #57: [Events] Add is_public and is_open columns to calendar event.
1. Write a migration to add `is_public` and `is_open` boolean columns to `calendar_events` table with default value as `false`. 
2. Update the sequelize model accordingly
3. Update CalendarEvent graphql type
4. Update the AuthorizationPolicy to allow update of the columns to organizer and admins and allow read to everyone.
5. Update CalendarEventCreate and CalendarEventUpdate mutations to support the new columns
6. Integrate the new columns in the create, view and edit event modals. In the create and edit modals have them as checkbox fields at the bottom named 'Public' and 'Open'. In the view modal have them as Badges of primary color at the bottom.
7. Update the CalendarEventRequest mutation to allow creating an invite request only if `is_open` is true for the event.
8. In the view event modal show request invite button only when `is_open` for the event is true.

Please maintain the code quality and naming conventions.

### PR DESCRIPTION
**Pull request by coderrsid**
## Fixes #57

- [x] Write a migration to add is_public and is_open boolean columns to calendar_events table with default value as false.
- [x] Update the sequelize model accordingly
- [x] Update CalendarEvent graphql type
- [x] Update the AuthorizationPolicy to allow update of the columns to organizer and admins and allow read to everyone.
- [x] Update CalendarEventCreate and CalendarEventUpdate mutations to support the new columns
- [x] Integrate the new columns in the create, view and edit event modals. In the create and edit modals have them as checkbox fields at the bottom named 'Public' and 'Open'. In the view modal have them as Badges of primary color at the bottom.
- [x] Update the CalendarEventRequest mutation to allow creating an invite request only if is_open is true for the event.
- [x] In the view event modal show request invite button only when is_open for the event is true.

To be added , if approved: 

- [x] Add the is_public and is_open attributes in API/services/CalendarEvent/Form?
- [x]  attribute is_open and is_public should be added to EventFragment GraphQL

### CODE REVIEW

File: api/graphql/mutation/calendarEventCreate/type.graphql [hunk 1]
```diff
@@ -5,6 +5,8 @@ input CalendarEventCreateInput {
   start_at: DateTime!
   end_at: DateTime!
   location: String
+  is_open: Boolean!
```
Comment (0xnmn) [inline api/graphql/mutation/calendarEventCreate/type.graphql]: No need to have them as required fields. By default they are both false.

File: api/graphql/mutation/calendarEventRequest/index.js [hunk 2]
```diff
@@ -21,7 +21,7 @@ class CalendarEventRequest extends BaseMutationResolver {
       },
     });
 
-    return form.create();
+    return event?.is_open && form.create();
```
Comment (0xnmn) [inline api/graphql/mutation/calendarEventRequest/index.js]: No need to check for is_open here. 

Move the validation logic from [CalendarEventInviteForm](https://github.com/coding-blocks/gondor/blob/master/api/services/CalendarEventInvite/Form.js#L15) to [AuthorizationPolicy](https://github.com/coding-blocks/gondor/blob/master/api/services/AuthorizationPolicy/index.js#L123). And also add the is_open check in that function.

Study this [lib](https://github.com/hereisnaman/auth-policy) to understand how auth policy works.

File: api/migrations/20200331220053-create-calendar-event.js [hunk 3]
```diff
@@ -46,6 +46,16 @@ module.exports = {
         allowNull: false,
         type: Sequelize.DATE,
       },
+      is_public: {
```
Comment (0xnmn) [inline api/migrations/20200331220053-create-calendar-event.js]: You have to create a new migration and add the columns there, this is completely wrong.

File: api/models/calendar_event.js [hunk 4]
```diff
@@ -43,6 +43,14 @@ module.exports = (sequelize, DataTypes) => {
...
         allowNull: false,
       },
+      is_public: {
+        type: DataTypes.BOOLEAN,
+        allowNull: false,
+      },
+      is_open: {
+        type: DataTypes.BOOLEAN,
+        allowNull: false,
+      },
```
Comment (0xnmn) [inline api/models/calendar_event.js]: Mention default value here.

File: app/modals/ViewEvent/Content/index.js [hunk 5]
```diff
@@ -124,6 +131,16 @@ const EventContent = memo(({ event, onClose }) => {
             <strong>Location:</strong> {event.location}
           </p>
         )}
+        {event.is_open && (
+          <>
```
Comment (0xnmn) [inline app/modals/ViewEvent/Content/index.js]: no need to use fragments

File: app/modals/ViewEvent/Content/index.js [hunk 6]
```diff
@@ -124,6 +131,16 @@ const EventContent = memo(({ event, onClose }) => {
...
+            <Badge color="primary">Open</Badge>
+          </>
+        )}
+        {event.is_public && (
+          <>
```
Comment (0xnmn) [inline app/modals/ViewEvent/Content/index.js]: no need to use fragments

File: app/modals/ViewEvent/Content/index.js [hunk 7]
```diff
@@ -124,6 +131,16 @@ const EventContent = memo(({ event, onClose }) => {
...
+            <Badge color="primary">Public</Badge>
```
Comment (0xnmn) [inline app/modals/ViewEvent/Content/index.js]: make it pill

File: app/modals/ViewEvent/Content/index.js [hunk 8]
```diff
@@ -124,6 +131,16 @@ const EventContent = memo(({ event, onClose }) => {
             <strong>Location:</strong> {event.location}
           </p>
         )}
+        {event.is_open && (
+          <>
+            <Badge color="primary">Open</Badge>
```
Comment (0xnmn) [inline app/modals/ViewEvent/Content/index.js]: make it pill

**Review by 0xnmn — CHANGES_REQUESTED** [review]

File: api/migrations/20200331220053-create-calendar-event.js (hunk 3 above)
Comment (coderrsid) [reply api/migrations/20200331220053-create-calendar-event.js]: Ohh sure.. What should i name the new migration file?

Comment (0xnmn) [reply api/migrations/20200331220053-create-calendar-event.js]: add_is_public_and_is_open_to_calendar_event

File: api/graphql/mutation/calendarEventRequest/index.js (hunk 2 above)
Comment (coderrsid) [reply api/graphql/mutation/calendarEventRequest/index.js]: Can you please explain a little more what i am supposed to do here?

File: api/graphql/mutation/calendarEventCreate/type.graphql (hunk 1 above)
Comment (coderrsid) [reply api/graphql/mutation/calendarEventCreate/type.graphql]: Done!

File: api/migrations/20200331220053-create-calendar-event.js (hunk 3 above)
Comment (coderrsid) [reply api/migrations/20200331220053-create-calendar-event.js]: Done!

File: api/models/calendar_event.js (hunk 4 above)
Comment (coderrsid) [reply api/models/calendar_event.js]: Done!

File: app/modals/ViewEvent/Content/index.js (hunk 5 above)
Comment (coderrsid) [reply app/modals/ViewEvent/Content/index.js]: Removed!

File: app/modals/ViewEvent/Content/index.js (hunk 6 above)
Comment (coderrsid) [reply app/modals/ViewEvent/Content/index.js]: Removed!

File: app/modals/ViewEvent/Content/index.js (hunk 8 above)
Comment (coderrsid) [reply app/modals/ViewEvent/Content/index.js]: Done!

File: app/modals/ViewEvent/Content/index.js (hunk 7 above)
Comment (coderrsid) [reply app/modals/ViewEvent/Content/index.js]: Done!

**Comment by coderrsid** [comment]
Can you review the migration file please, @hereisnaman

**Comment by 0xnmn** [comment]
@coderrsid The migration looks good. Now for the authorization. In the AuthPolicy for the action `calendarEventInvite:status:update` check if value.status === 'Requested' && entity.event.is_open.

There are other conditions there too, don't change them.

**Comment by coderrsid** [comment]
> @coderrsid The migration looks good. Now for the authorization. In the AuthPolicy for the action `calendarEventInvite:status:update` check if value.status === 'Requested' && entity.event.is_open.
> 
> There are other conditions there too, don't change them.

So the `value.status === 'Requested' && entity.event.is_open.` should be in both conditions - `requested` and `!requested`. Or i should just wrap this condition around the 2 conditions in `calendarEventInvite:status:update`

**Comment by coderrsid** [comment]
> @coderrsid The migration looks good. Now for the authorization. In the AuthPolicy for the action `calendarEventInvite:status:update` check if value.status === 'Requested' && entity.event.is_open.
> 
> There are other conditions there too, don't change them.

@hereisnaman Through this condition we are gonna check whether if status by viewer is requested and if the event is open, viewer should be atleast `isMember` for event invite, am i understanding it right?

**Comment by 0xnmn** [comment]
> > @coderrsid The migration looks good. Now for the authorization. In the AuthPolicy for the action `calendarEventInvite:status:update` check if value.status === 'Requested' && entity.event.is_open.
> > There are other conditions there too, don't change them.
> 
> @hereisnaman Through this condition we are gonna check whether if status by viewer is requested and if the event is open, viewer should be atleast `isMember` for event invite, am i understanding it right?

If the viewer has requested and the event is open then allow everyone. But if the event is not open then allow only members (that includes admins too when you use isMember func)

**Comment by coderrsid** [comment]
> If the viewer has requested and the event is open then allow everyone. But if the event is not open then allow only members (that includes admins too when you use isMember func)

I guess the last commit should do the work as you suggested. If not please do tell me the changes. Thanks

**Comment by 0xnmn** [comment]
> > If the viewer has requested and the event is open then allow everyone. But if the event is not open then allow only members (that includes admins too when you use isMember func)
> 
> I guess the last commit should do the work as you suggested. If not please do tell me the changes. Thanks

I don't think you have even tested it locally, because it doesn't work. The PR will be closed if this behavior is repeated.

**Comment by coderrsid** [comment]
> I don't think you have even tested it locally, because it doesn't work. The PR will be closed if this behavior is repeated.

@hereisnaman  I am so sorry, this'd not be repeated. I tried to test but now i see error in log that
`original: error: column calendar_event.is_public does not exist` in the SQL query
image

Is their a problem with migration or it's just not executing?

**Comment by 0xnmn** [comment]
> > I don't think you have even tested it locally, because it doesn't work. The PR will be closed if this behavior is repeated.
> 
> @hereisnaman I am so sorry, this'd not be repeated. I tried to test but now i see error in log that
> `original: error: column calendar_event.is_public does not exist` in the SQL query
> image
> 
> Is their a problem with migration or it's just not executing?

Have you run the migration?

**Comment by coderrsid** [comment]
> > > I don't think you have even tested it locally, because it doesn't work. The PR will be closed if this behavior is repeated.
> > 
> > 
> > @hereisnaman I am so sorry, this'd not be repeated. I tried to test but now i see error in log that
> > `original: error: column calendar_event.is_public does not exist` in the SQL query
> > image
> > Is their a problem with migration or it's just not executing?
> 
> Have you run the migration?

Can you guide me how to do that please?

**Comment by coderrsid** [comment]
@hereisnaman  why am i not able to add event or do any CRUD to event. Even though AuthPolicy asks for member, how am i supposed to test the feature? 
image

It shows 
image

**Comment by 0xnmn** [comment]
> @hereisnaman why am i not able to add event or do any CRUD to event. Even though AuthPolicy asks for member, how am i supposed to test the feature?
> image
> 
> It shows
> image

The problem is for you to solve.

**Comment by coderrsid** [comment]
> The problem is for you to solve.

I know mahn. But this is what i am facing from the starting and i asked you event at that time in an issue but you said read auth_policy, i couldn't understand it but after you suggested i read auth policy docs but not only me many of us are not able to solve this.

**Comment by 0xnmn** [comment]
> > The problem is for you to solve.
> 
> I know mahn. But this is what i am facing from the starting and i asked you event at that time in an issue but you said read auth_policy, i couldn't understand it but after you suggested i read auth policy docs but not only me many of us are not able to solve this.

What is the role of your user?

**Comment by coderrsid** [comment]
> > > The problem is for you to solve.
> > 
> > 
> > I know mahn. But this is what i am facing from the starting and i asked you event at that time in an issue but you said read auth_policy, i couldn't understand it but after you suggested i read auth policy docs but not only me many of us are not able to solve this.
> 
> What is the role of your user?

I guess developer. But i think isSelf should work?

**Comment by 0xnmn** [comment]
> > > > The problem is for you to solve.
> > > 
> > > 
> > > I know mahn. But this is what i am facing from the starting and i asked you event at that time in an issue but you said read auth_policy, i couldn't understand it but after you suggested i read auth policy docs but not only me many of us are not able to solve this.
> > 
> > 
> > What is the role of your user?
> 
> I guess developer. But i think isSelf should work?

Read the readme and convert your account to Admin or Member. The authorization policy clearly states that.

**Comment by 0xnmn** [comment]
@coderrsid The following is the correct AuthPolicy:
Screenshot 2020-05-21 at 2 43 39 AM


Other than this, there are more issue including:
- broken css for checkbox buttons. 
- state for is_public and is_open is not intialized with saved value in edit form.
- no margin between pills in view modal. give ml-1 to the right pill

Screenshot 2020-05-21 at 2 37 01 AM
Screenshot 2020-05-21 at 2 37 14 AM

**Comment by coderrsid** [comment]
Thanks. I figured it out last night only. Now i guess it is working. Don't worry, i am adding a GIF too.
Please tell me if you think any more functionality or styling should be modified.
Screen-Recording-2020-05-21-at-9

File: app/components/Events/Form.js [hunk 9]
```diff
@@ -32,6 +36,8 @@ const EventForm = memo(
   }) => {
     const viewer = useViewer();
 
+    console.log(isOpen, isPublic);
```
Comment (0xnmn) [inline app/components/Events/Form.js]: remove log

**Review by 0xnmn — APPROVED** [review]

**Comment by 0xnmn** [comment]
@coderrsid good work. Just remove the log. I'll merge.

**Comment by coderrsid** [comment]
> @coderrsid good work. Just remove the log. I'll merge.

Thanks, you've helped me quite a lot. I know right now i don't have that much knowledge of gondor's codebase but i can assure you i'll get the hang of it soon and will perform better. 😄

**Pull request merged**